couchbase/io/streams.hxx

```diff
@@ -72,71 +81,77 @@ class stream_impl
 class plain_stream_impl : public stream_impl
 {
   private:
-    asio::ip::tcp::socket stream_;
+    std::shared_ptr<asio::ip::tcp::socket> stream_;
 
   public:
     explicit plain_stream_impl(asio::io_context& ctx)
       : stream_impl(ctx, false)
-      , stream_(strand_)
+      , stream_(std::make_unique<asio::ip::tcp::socket>(strand_))
     {
     }
 
-    [[nodiscard]] bool is_open() const override
-    {
-        return stream_.is_open();
-    }
-
     [[nodiscard]] asio::ip::tcp::endpoint local_endpoint() const override
     {
         std::error_code ec;
-        auto res = stream_.local_endpoint(ec);
+        auto res = stream_->local_endpoint(ec);
         if (ec) {
             return {};
         }
         return res;
     }
 
-    void close() override
+    void close(std::function<void(std::error_code)>&& handler) override
     {
-        stream_.close();
+        open_ = false;
+        return asio::post(strand_, [stream = stream_, h = std::move(handler)]() {
+            asio::error_code ec{};
+            stream->shutdown(asio::socket_base::shutdown_both, ec);
+            stream->close(ec);
+            h(ec);
+        });
     }
 
     void reopen() override
     {
-        id_ = uuid::to_string(uuid::random());
-        asio::error_code ec{};
-        stream_.shutdown(asio::socket_base::shutdown_both, ec);
-        stream_.close(ec);
-        stream_ = asio::ip::tcp::socket(strand_);
+        return close([this](std::error_code) {
+            id_ = uuid::to_string(uuid::random());
+            stream_ = std::make_shared<asio::ip::tcp::socket>(strand_);
```

Comment:
There is likely still race conditions here as while the close handler might execute on the strand, there are a bunch of places where the the fields of the stream (like id_ and stream_) are being accessed off the strand.  The shared_ptr holding the stream is not safely accessible from off the strand since it can change on-strand.  id_ is similar.  That being said, again, I think it's worthwhile to merge this and address these issues later.

Reply:
https://issues.couchbase.com/browse/CXXCBC-74